diff --git a/src/Controller/ImageStyleDownloadController.php b/src/Controller/ImageStyleDownloadController.php
index c8c9d67..9c33d76 100644
--- a/src/Controller/ImageStyleDownloadController.php
+++ b/src/Controller/ImageStyleDownloadController.php
@@ -33,7 +33,7 @@ class ImageStyleDownloadController extends CoreImageStyleDownloadController {
   /**
    * {@inheritdoc}
    */
-  public function deliver(Request $request, $scheme, ImageStyleInterface $image_style) {
+  public function deliver(Request $request, $scheme, ImageStyleInterface $image_style, string $required_derivative_scheme) {
     $target = $request->query->get('file');
     if (!$target) {
       throw new NotFoundHttpException();
@@ -46,7 +46,7 @@ class ImageStyleDownloadController extends CoreImageStyleDownloadController {
       if ($source_uri = $this->lookupSourceImage($image_uri)) {
         // Replace webp image with source image and call parent:deliver().
         $request->query->set('file', str_replace($scheme . '://', '', $source_uri));
-        $source_response = parent::deliver($request, $scheme, $image_style);
+        $source_response = parent::deliver($request, $scheme, $image_style, $required_derivative_scheme);
         $derivative_uri = $image_style->buildUri($image_uri);
         // If parent:deliver() returns BinaryFileResponse, we'll replace
         // the BinaryFileResponse with one containing the .webp image
@@ -71,7 +71,7 @@ class ImageStyleDownloadController extends CoreImageStyleDownloadController {
       throw new NotFoundHttpException();
     }
     else {
-      return parent::deliver($request, $scheme, $image_style);
+      return parent::deliver($request, $scheme, $image_style, $required_derivative_scheme);
     }
   }
 
