diff --git a/src/PathautoGenerator.php b/src/PathautoGenerator.php
index 0000000..0000000 100644
--- a/src/PathautoGenerator.php
+++ b/src/PathautoGenerator.php
@@ -315,20 +315,22 @@ class PathautoGenerator implements PathautoGeneratorInterface {
    * {@inheritdoc}
    */
   public function getPatternByEntity(EntityInterface $entity) {
-    $langcode = $entity->language()->getId();
-    if (!isset($this->patterns[$entity->getEntityTypeId()][$entity->id()][$langcode])) {
+    $entity_type_id = $entity->getEntityTypeId() ?? '';
+    $entity_id = $entity->id() ?? '';
+    $langcode = $entity->language()->getId() ?? '';
+    if (!isset($this->patterns[$entity_type_id][$entity_id][$langcode])) {
       foreach ($this->getPatternByEntityType($entity->getEntityTypeId()) as $pattern) {
         if ($pattern->applies($entity)) {
-          $this->patterns[$entity->getEntityTypeId()][$entity->id()][$langcode] = $pattern;
+          $this->patterns[$entity_type_id][$entity_id][$langcode] = $pattern;
           break;
         }
       }
       // If still not set.
-      if (!isset($this->patterns[$entity->getEntityTypeId()][$entity->id()][$langcode])) {
-        $this->patterns[$entity->getEntityTypeId()][$entity->id()][$langcode] = NULL;
+      if (!isset($this->patterns[$entity_type_id][$entity_id][$langcode])) {
+        $this->patterns[$entity_type_id][$entity_id][$langcode] = NULL;
       }
     }
-    return $this->patterns[$entity->getEntityTypeId()][$entity->id()][$langcode];
+    return $this->patterns[$entity_type_id][$entity_id][$langcode];
   }
 
   /**
