diff --git a/src/PathautoGenerator.php b/src/PathautoGenerator.php
index 8d60e57..5d91697 100644
--- a/src/PathautoGenerator.php
+++ b/src/PathautoGenerator.php
@@ -343,7 +343,7 @@ class PathautoGenerator implements PathautoGeneratorInterface {
    */
   public function getPatternByEntity(EntityInterface $entity) {
     $langcode = $entity->language()->getId();
-    if (!isset($this->patterns[$entity->getEntityTypeId()][$entity->id()][$langcode])) {
+    if ($entity->id() === NULL || !isset($this->patterns[$entity->getEntityTypeId()][$entity->id()][$langcode])) {
       foreach ($this->getPatternByEntityType($entity->getEntityTypeId()) as $pattern) {
         if ($pattern->applies($entity)) {
           // Fetch the translated pattern string if a language-specific config
@@ -356,7 +356,7 @@ class PathautoGenerator implements PathautoGeneratorInterface {
               $pattern->setPattern($translated);
             }
           }
-          if ($entity->isNew()) {
+          if ($entity->id() === NULL || $entity->isNew()) {
             return $pattern;
           }
           $this->patterns[$entity->getEntityTypeId()][$entity->id()][$langcode] = $pattern;
@@ -364,11 +364,11 @@ class PathautoGenerator implements PathautoGeneratorInterface {
         }
       }
       // If still not set.
-      if (!isset($this->patterns[$entity->getEntityTypeId()][$entity->id()][$langcode])) {
+      if ($entity->id() !== NULL && !isset($this->patterns[$entity->getEntityTypeId()][$entity->id()][$langcode])) {
         $this->patterns[$entity->getEntityTypeId()][$entity->id()][$langcode] = NULL;
       }
     }
-    return $this->patterns[$entity->getEntityTypeId()][$entity->id()][$langcode];
+    return $entity->id() !== NULL ? $this->patterns[$entity->getEntityTypeId()][$entity->id()][$langcode] : NULL;
   }
 
   /**
