diff --git a/src/AuthManager/OAuth2Manager.php b/src/AuthManager/OAuth2Manager.php
index 6cea9b0..ff845a3 100755
--- a/src/AuthManager/OAuth2Manager.php
+++ b/src/AuthManager/OAuth2Manager.php
@@ -73,7 +73,7 @@ abstract class OAuth2Manager implements OAuth2ManagerInterface {
    */
   public function __construct(ImmutableConfig $settings,
                               LoggerChannelFactoryInterface $logger_factory,
-                              Request $request = NULL) {
+                              ?Request $request = NULL) {
 
     $this->settings = $settings;
     $this->loggerFactory = $logger_factory;
